[FX] Proxyable classes

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56737 [FX][WIP] Proxyable classes**

This patch adds a metaclass, `fx.ProxyableClassMeta`, that enables construction of objects that use this metaclass to be traced in symbolic tracing. More details can be found in the docstring

TODOs:

- [x] Ensure that serialization is okay, both for serialization of the identifier for constructor targets as well as serializing interned instances. UPDATE: seems to work  https://gist.github.com/jamesr66a/5623e9f92300dd513189270bcf4f8b90
- [x] Write documentation
- [x] UX study
- [x] Validate with users

Differential Revision: [D27953073](https://our.internmc.facebook.com/intern/diff/D27953073)

Commit:
[FX][WIP] Proxyable classes

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Proxyable classes"


TODOs:

- [ ] Ensure that serialization is okay, both for serialization of the identifier for constructor targets as well as serializing interned instances
- [ ] Write documentation
- [ ] UX study
- [ ] Validate with users

Differential Revision: [D27953073](https://our.internmc.facebook.com/intern/diff/D27953073)

[ghstack-poisoned]

wanchaol reviewed (commented):
Thanks for this feature! When experimenting this, I think there're almost no issues except one: when the custom ProxyNode has property and we access it during tracing, it seems we didn't handle that well. Example:
```python
class TestMeta(metaclass=torch.fx.ProxyableClassMeta):
    def __init__(self, keys: List[str]):
        self._keys = keys

    @property
    def keyes(self) -> List[str]:
        return self._keys


class TestModule(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, test_input: TestMeta) -> int:
        return len(test_input.keys)

test_mod = TestModule()

gm = torch.fx.symbolic_trace(
    test_mod,
)
print(gm.code)
```
This crashes with error:
`TypeError: object of type 'method' has no len()`

Do we need to do some special handling for class properties?

Commit:
Update on "[FX][WIP] Proxyable classes"


TODOs:

- [ ] Ensure that serialization is okay, both for serialization of the identifier for constructor targets as well as serializing interned instances
- [ ] Write documentation
- [ ] UX study
- [ ] Validate with users

Differential Revision: [D27953073](https://our.internmc.facebook.com/intern/diff/D27953073)

[ghstack-poisoned]

jamesr66a commented:
@jamesr66a has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27953073).

jamesr66a commented:
Hey @wanchaol,

It looks like that is happening because there is a conflict between your class's definition and the proxy handling for `**` unpacking: https://github.com/pytorch/pytorch/blob/a13718b69f7df3caf12e5056bccea118eb5534a1/torch/fx/proxy.py#L140

This works if you name the property something other than `keys`:

```
import torch
import torch.fx
from typing import List

torch.fx.wrap('len')

class TestMeta(metaclass=torch.fx.ProxyableClassMeta):
    def __init__(self, keys: List[str]):
        self._keys = keys

    @property
    def my_keys(self) -> List[str]:
        return self._keys


class TestModule(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, test_input: TestMeta) -> int:
        return len(test_input.my_keys)

test_mod = TestModule()

gm = torch.fx.symbolic_trace(
    test_mod,
)
print(gm.code)
```


```

def forward(self, test_input : __main___TestMeta) -> int:
    getattr_1 = test_input.my_keys;  test_input = None
    len_1 = len(getattr_1);  getattr_1 = None
    return len_1
```

Does that work for you?

wanchaol commented:
> It looks like that is happening because there is a conflict between your class's definition and the proxy handling for `**` unpacking:
> 
> https://github.com/pytorch/pytorch/blob/a13718b69f7df3caf12e5056bccea118eb5534a1/torch/fx/proxy.py#L140

Hmmm interesting to know. Does that mean that user defined classes couldn't defined "keys" as their attribute? I think "keys" is a commonly used words for things like dictionary and customized tensor with keys, it might be weird if it's marked as a builtin method/attribute and user class couldn't use it. 

Proxy object can be called with `**` but it only need to have `keys` when it's things like `OrderedDict` or things similar? If the ProxyObject is not a dict like object, this should not be there as an attribute iiuc. Do we have a way to hide this attribute or make this attribute be as `__keys__` instead as it's really only the unpacking details?

Commit:
Update on "[FX][WIP] Proxyable classes"


TODOs:

- [ ] Ensure that serialization is okay, both for serialization of the identifier for constructor targets as well as serializing interned instances
- [ ] Write documentation
- [x] UX study
- [x] Validate with users

Differential Revision: [D27953073](https://our.internmc.facebook.com/intern/diff/D27953073)

[ghstack-poisoned]

jamesr66a commented:
@jamesr66a has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27953073).

Chillee commented on torch/fx/symbolic_trace.py:
```diff
@@ -23,6 +23,75 @@
+    tracing.
+    """
+    def __init__(cls, name, bases, attrs):
+        _proxyable_classes.setdefault(cls)
```
why is this not a set instead of a dict?

jamesr66a replied:
Sets are unordered but dicts are ordered. Dict with dont-care value is a pattern used for "ordered set" semantics. I prefer to use ordered sets because it makes things deterministic (so we're not chasing down weird errors either in tests or in user code)

Chillee reviewed (approved):
LGTM

jamesr66a commented on torch/fx/symbolic_trace.py:
```diff
@@ -225,6 +294,22 @@ def create_arg(self, a: Any) -> 'Argument':
+            # This is an instance of a proxyable class for which we did not
+            # witness its construction. Intern this as a constant attribute
+
+            # TODO: binary search
```
![image](https://user-images.githubusercontent.com/4685384/118309740-8a203400-b4a2-11eb-9c13-03f2cfa5873b.png)
